Change how resources expose urls to the dashboard

davidfowl opened the pull request:
- The resource server used to have a services and endpoints split, that is now a single object called a url which has a name, a full url (uri format) and a bool for if that endpoint is internal (or an implementation detail). Internal urls only show up in the details pane.
- Clean up the way we extract urls from the dcp information by mapping the app model resource and endpoint within each service resource in DCP.
- Updated tests

Fixes https://github.com/dotnet/aspire/issues/3078

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3103)

davidfowl committed:
Change how resources expose urls to the dashboard
- The resource server used to have a services and
endpoints split, that is now a single object called a url
which has a name, a full url (uri format) and a bool for
if that endpoint is internal (or an implementation detail).
- Clean up the way we extract urls from the dcp information by
mapping the app model resource and endpoint within each service
resource in DCP.
- Updated tests

davidfowl committed:
Fixed error

davidfowl committed:
Added tests

drewnoakes commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -599,58 +556,83 @@ string CombineUrls(string url, string launchUrl)
-        var services = ImmutableArray.CreateBuilder<(string Name, string? AllocatedAddress, int? AllocatedPort)>();
         var name = resource.Metadata.Name;
 
+        var urls = ImmutableArray.CreateBuilder<(string Name, string Url, bool IsInternal)>();
```
nit: this allocation could be made lazy by making this nullable, so we don't allocate an object when there are no URLs.

drewnoakes commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -599,58 +556,83 @@ string CombineUrls(string url, string launchUrl)
+                        var url = CombineUrls(ep.Url, launchUrl);
 
-                if (launchProfile is not null)
+                        urls.Add(new(ep.EndpointName, url, false));
```
nit: consider naming literals argument for context:

```suggestion
                        urls.Add(new(ep.EndpointName, url, IsInternal: false));
```

davidfowl replied:
Not possible with the value tuple. I need to create real types for this API and just merge it with ResourceSnapshot

JamesNK replied:
Records > tuples IMO

davidfowl replied:
Ya'll made me do all the work I didn't want to do. It's much cleaner though.

davidfowl replied:
Done

drewnoakes commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -599,58 +556,83 @@ string CombineUrls(string url, string launchUrl)
-                    _ = bool.TryParse(launchProfile, out isFromLaunchProfile);
+                    if (ep.IsAllocated)
+                    {
+                        urls.Add(new(ep.EndpointName, ep.Url, false));
```
nit: consider naming literals argument for context:

```suggestion
                        urls.Add(new(ep.EndpointName, ep.Url, IsInternal: false));
```

davidfowl replied:
Same as above.

davidfowl replied:
Done in the latest

drewnoakes commented on src/Aspire.Hosting/Dcp/ApplicationExecutor.cs:
```diff
@@ -599,58 +556,83 @@ string CombineUrls(string url, string launchUrl)
                 {
-                    services.Add(new(service.Metadata.Name, service.AllocatedAddress, service.AllocatedPort));
+                    var endpointString = $"{ep.Scheme}://{endpoint.Spec.Address}:{endpoint.Spec.Port}";
+                    urls.Add(new($"{ep.EndpointName}-listen-port", endpointString, true));
```
nit: consider naming literals argument for context:

```suggestion
                    urls.Add(new(Name: $"{ep.EndpointName}-listen-port", endpointString, IsInternal: true));
```

davidfowl replied:
Done

drewnoakes reviewed (approved):
Much nicer.

Love to see this too:

![image](https://github.com/dotnet/aspire/assets/350947/3b7fffe7-7213-43ee-a00d-c64c3b8a7163)

davidfowl commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -139,6 +149,9 @@ message Resource {
     repeated ResourceProperty properties = 12;
+
+    repeated Url urls = 13;
+    bool expect_urls = 14;
```
I thought about removing it but the ux of none vs starting was something I didn’t want to change

davidfowl commented on src/Aspire.Dashboard/Components/ResourcesGridColumns/EndpointsColumnDisplay.razor:
```diff
@@ -6,9 +6,9 @@
     string? additionalMessage = null;
-    if (Resource.Endpoints.Length == 0 && (Resource.State == ResourceStates.FinishedState || Resource.ExpectedEndpointsCount == 0))
+    if (Resource.Urls.Length == 0 && (Resource.State == ResourceStates.FinishedState || Resource.ExpectUrls is false))
     {
-        if (Resource.Services.Length == 0)
+        if (Resource.ExpectUrls is false)
```
Will use !

davidfowl committed:
PR feedback and clean up
- Expose ExpectUrls on CustomResourceSnapshot
- Added a comment to the proto file

JamesNK commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -127,8 +133,12 @@ message Resource {
     optional string state = 5;
     optional google.protobuf.Timestamp created_at = 6;
     repeated EnvironmentVariable environment = 7;
+    // Deprecated
     optional int32 expected_endpoints_count = 8;
```
https://protobuf.dev/programming-guides/proto3/#options

You can add deprecated option to a field. To be honest I'm not exactly sure what impact it has in C# (it might add obsolete attribute, or it might do nothing). But give it a try :smile:

e.g
```proto
optional int32 expected_endpoints_count = 8 [deprecated = true];
```

davidfowl replied:
Done

JamesNK commented:
I got an error when I ran TestShop on this branch:

```
System.Net.Http.HttpRequestException: No such host is known. (catalogservice:443)
 ---> System.Net.Sockets.SocketException (11001): No such host is known.
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|285_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(String host, Int32 port, HttpRequestMessage initialRequest, Boolean async, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(String host, Int32 port, HttpRequestMessage initialRequest, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(QueueItem queueItem)
   at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.HttpConnectionWaiter`1.WaitForConnectionWithTelemetryAsync(HttpRequestMessage request, HttpConnectionPool pool, Boolean async, CancellationToken requestCancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.<SendCoreAsync>g__Core|5_0(HttpRequestMessage request, Boolean useAsync, CancellationToken cancellationToken)
   at Microsoft.Extensions.ServiceDiscovery.Http.ResolvingHttpDelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in C:\Development\Source\aspire\src\Microsoft.Extensions.ServiceDiscovery\Http\ResolvingHttpDelegatingHandler.cs:line 59
   at Microsoft.Extensions.Http.Resilience.ResilienceHandler.<>c.<<SendAsync>b__3_0>d.MoveNext()
--- End of stack trace from previous location ---
   at Polly.Outcome`1.ThrowIfException()
   at Microsoft.Extensions.Http.Resilience.ResilienceHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.<SendCoreAsync>g__Core|5_0(HttpRequestMessage request, Boolean useAsync, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at System.Net.Http.Json.HttpClientJsonExtensions.<FromJsonAsyncCore>g__Core|12_0[TValue,TJsonOptions](HttpClient client, Task`1 responseTask, Boolean usingResponseHeadersRead, CancellationTokenSource linkedCTS, Func`4 deserializeMethod, TJsonOptions jsonOptions, CancellationToken cancellationToken)
   at MyFrontend.Components.Pages.Home.OnInitializedAsync() in C:\Development\Source\aspire\playground\TestShop\MyFrontend\Components\Pages\Home.razor:line 77
   at Microsoft.AspNetCore.Components.ComponentBase.RunInitAndSetParametersAsync()
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle, ComponentState owningComponentState)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle, ComponentState owningComponentState)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.<WaitForQuiescence>g__ProcessAsynchronousWork|54_0()
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.WaitForQuiescence()
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.RenderRootComponentAsync(Int32 componentId, ParameterView initialParameters)
   at Microsoft.AspNetCore.Components.Endpoints.EndpointHtmlRenderer.SendStreamingUpdatesAsync(HttpContext httpContext, Task untilTaskCompleted, TextWriter writer)
```

JamesNK commented on src/Aspire.Dashboard/Model/ResourceOutgoingPeerResolver.cs:
```diff
@@ -96,9 +96,13 @@ bool TryMatchResourceAddress(string value, [NotNullWhen(true)] out string? name)
+                foreach (var service in resource.Urls)
                 {
-                    if (string.Equals(service.AddressAndPort, value, StringComparison.OrdinalIgnoreCase))
+                    var uri = new Uri(service.Url);
```
`Uri.TryCreate` is used ealier. Should be tolerant of a badly formed Url here as well.

davidfowl replied:
Done, stored the uri in the view model

JamesNK commented on src/Aspire.Dashboard/Model/ResourceViewModel.cs:
```diff
@@ -92,37 +91,20 @@ public EnvironmentVariableViewModel(string name, string? value, bool fromSpec)
-    public int? AllocatedPort { get; }
-
-    public string AddressAndPort { get; }
+    public string Url { get; }
```
Store a parsed Uri in the VM?

davidfowl replied:
Done, stored the uri in the view model

JamesNK commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -108,6 +108,12 @@ message Service {
+message Url {
+    string name = 1;
+    string full_url = 2;
+    bool is_internal = 3;
```
Add a comment for what this means

davidfowl replied:
Done

JamesNK commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -127,8 +133,12 @@ message Resource {
     optional string state = 5;
     optional google.protobuf.Timestamp created_at = 6;
     repeated EnvironmentVariable environment = 7;
+    // Deprecated
     optional int32 expected_endpoints_count = 8;
+
+    // Deprecated
     repeated Endpoint endpoints = 9;
+    // Deprecated
     repeated Service services = 10;
```
Let me know what the word is with protocol changes. I want one for [logging](https://github.com/dotnet/aspire/pull/3075) (add line number to log messages).

JamesNK commented on src/Aspire.Hosting/Dashboard/proto/resource_service.proto:
```diff
@@ -139,6 +149,13 @@ message Resource {
+    // Returns true if there will ever be any urls. Since these updates are being sent as the state
+    // of resources change, it is possible to for urls to be empty but expect_urls to be true.
+    // If urls is non-empty, expect_urls must be true.
+    bool expect_urls = 14;
```
A `Url` isn't present in the urls collection until it is started/ready, right? And that's why this property is required?

Is there a reason we couldn't instead have all the named urls in the urls collection no matter what, and then have a state on url that indicates whether it's ready?

davidfowl replied:
It’s not required. We can remove it, but the dashboard will go from None -> URLs instead of Starting -> URLs. I don’t think it’s worth it having this complexity. I was just trying to preserve the current state

davidfowl replied:
Removed

davidfowl committed:
PR feedback
- Store parsed uri in the ResourceViewModel
- Remove ExpectUrls
- Mark fields as deprecated in the proto file

davidfowl committed:
Moved from Tuples to records for CustomResourceSnapshot
